File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {


        vector<vector<int>> adjacent(numCourses);

        vector<int> visited(numCourses, 0);

        for (auto& edges : prerequisites) {
            int u = edges[0];
            int v = edges[1];
            adjacent[v].push_back(u);

        }

        for (int i = 0; i < numCourses; i++) {
            if (visited[i] == 0 && !dfs(i, adjacent, visited))  return false;
        }

        return true;
        
    }


    bool dfs (int i, vector<vector<int>>& adjacent, vector<int>& visited) {
        if (visited[i] == 1) {
            return false;
        }

        if (visited[i] == 2) {
            return true;
        }

        visited[i] = 1;

        for (auto& edge : adjacent[i]) {
            if (!dfs(edge, adjacent, visited)) return false;
        }

        visited[i] = 2;

        return true;
    }
};

File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:

        preMap = {i:[] for i in range(numCourses)}

        visitSet = set()

        for pre, crs in prerequisites:
            preMap[pre].append(crs)
        
        def dfs(crs):
            if crs in visitSet:
                return False
            
            if preMap[crs] == []:
                return True
            
            visitSet.add(crs)
            for pre in preMap[crs]:
                if not dfs(pre):
                    return False
            
            preMap[crs] = []
            visitSet.remove(crs)
            return True
            
        for crs in range(numCourses):
            if not dfs(crs):
                return False
        
        return True
        

        